pattern/constraint/value/AssignmentConstraint.java

```diff
@@ -0,0 +1,624 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.vaticle.typedb.core.pattern.constraint.value;
+
+import com.vaticle.typedb.common.collection.Pair;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.encoding.Encoding;
+import com.vaticle.typedb.core.pattern.Conjunction;
+import com.vaticle.typedb.core.pattern.constraint.Constraint;
+import com.vaticle.typedb.core.pattern.variable.ThingVariable;
+import com.vaticle.typedb.core.pattern.variable.ValueVariable;
+import com.vaticle.typedb.core.pattern.variable.Variable;
+import com.vaticle.typedb.core.pattern.variable.VariableCloner;
+import com.vaticle.typedb.core.pattern.variable.VariableRegistry;
+import com.vaticle.typedb.core.traversal.GraphTraversal;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+import com.vaticle.typeql.lang.common.TypeQLToken;
+import com.vaticle.typeql.lang.pattern.constraint.ValueConstraint.Assignment.Expression.Operation.Op;
+import com.vaticle.typeql.lang.pattern.variable.BoundVariable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static com.vaticle.typedb.common.util.Objects.className;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Pattern.INVALID_CASTING;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class AssignmentConstraint extends ValueConstraint {
+
+    private final Expression expression;
+    private final Set<Variable> arguments;
+    private final int hash;
+
+    public AssignmentConstraint(ValueVariable owner, Expression expression, Set<Variable> arguments) {
+        super(owner, arguments);
+        this.expression = expression;
+        this.arguments = arguments;
+        this.hash = Objects.hash(this.owner, this.expression);
+    }
+
+    public static AssignmentConstraint of(ValueVariable valueVariable, com.vaticle.typeql.lang.pattern.constraint.ValueConstraint.Assignment constraint, VariableRegistry registry) {
+        Set<Variable> arguments = new HashSet<>();
+        for (BoundVariable v : constraint.variables()) {
+            if (v.isThing()) arguments.add(registry.register(v.asThing()));
+            else if (v.isValue()) arguments.add(registry.register(v.asValue()));
+            else throw TypeDBException.of(ILLEGAL_STATE);
+        }
```

Comment:
It's a bit different from our usual pattern to pre-register all the variables first, then retrieve them. However I don't think this approach is too bad in this context